tests/check/check_test.go

```diff
@@ -18,28 +18,38 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
 	healthv1pb "google.golang.org/grpc/health/grpc_health_v1"
-	"gopkg.in/yaml.v2"
+	"google.golang.org/grpc/status"
+	"gopkg.in/yaml.v3"
 )
 
 type checkTests struct {
-	Tests []checkTest
+	Tests []*checkTest
 }
 
 type checkTest struct {
 	Name   string
-	Stages []stage
+	Stages []*stage
 }
 
+// stage is a stage of a test. All stages will be run in a single store.
 type stage struct {
 	Model      string
 	Tuples     []*pb.TupleKey
-	Assertions []assertion
+	Assertions []*assertion
+	Validation *validation
```

Comment:
Adding `validation` here feels a bit meta. I think these test cases would be more uniform if we just added a `code` field to an assertion (default value of 0 will be `OK`). Then we can add validation test assertions whose expected code is a validation error code. 

This would make each test case logic the same and wouldn't mean there are metadata differences between individual test cases.

Reply:
For example,

```
  - name: validation_relation_not_in_model
    stages:
      - model: |
          type user

        assertions:
          - tuple:
              object: user:aardvark
              relation: viewer
              user: user:badger
            expectation: false
            code: 2009
```